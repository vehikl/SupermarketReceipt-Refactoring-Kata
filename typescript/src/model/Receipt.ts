import {Discount} from "./Discount"
import {ReceiptItem} from "./ReceiptItem"

export class Receipt {
    private readonly items: ReceiptItem[] = [];
    private readonly discounts: Discount[] = [];

    public constructor(items: Array<ReceiptItem>, discounts: Array<Discount>) {
        this.items = items;
        this.discounts = discounts;
    }

    public getTotalPrice(): number {
        let total = 0.0;
        for (let item of this.items) {
            total += item.totalPrice;
        }
        for (let discount of this.discounts) {
            total -= discount.getDiscountAmount();
        }
        return total;
    }

    public getItems(): ReceiptItem[] {
        return this.items;
    }

    public getDiscounts(): Discount[] {
        return this.discounts;
    }
}
