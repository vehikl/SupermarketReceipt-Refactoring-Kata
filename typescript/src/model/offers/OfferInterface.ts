import { Discount } from "../Discount"
import { ShoppingCart } from "../ShoppingCart";

interface OfferInterface {
  getDiscount(cart: ShoppingCart): Discount;
  applies(cart: ShoppingCart): boolean;
}

export default OfferInterface;