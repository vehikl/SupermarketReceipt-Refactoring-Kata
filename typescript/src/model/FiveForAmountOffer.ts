import { Product } from "./Product"
import { Discount } from "./Discount"
import { ShoppingCart } from "./ShoppingCart";
import OfferInterface from './OfferInterface';


export class FiveForAmountOffer implements OfferInterface {
  private minimumQuantityForOffer: number = 5;
  private product: Product;
  private unitPrice: number;
  private discountedPrice: number;

  public constructor(product: Product, unitPrice: number, discountedPrice: number) {
    this.product = product;
    this.unitPrice = unitPrice;
    this.discountedPrice = discountedPrice;
  }

  public getDiscount(cart: ShoppingCart) {
    const quantity = cart.getQuantityOf(this.product);
    const total = this.discountedPrice * Math.floor(quantity / this.minimumQuantityForOffer) + quantity % this.minimumQuantityForOffer * this.unitPrice;
    const discountAmount = this.unitPrice * quantity - total;
    return new Discount(this.product, `${this.minimumQuantityForOffer} for ${this.discountedPrice}`, discountAmount);
  }

  public applies(cart: ShoppingCart): boolean {
    return cart.getQuantityOf(this.product) >= this.minimumQuantityForOffer;
  }
}