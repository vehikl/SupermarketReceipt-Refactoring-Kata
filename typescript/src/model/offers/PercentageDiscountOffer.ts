import { ShoppingCart } from "../ShoppingCart";
import { Product } from "../Product";
import { Discount } from "../Discount";
import OfferInterface from './OfferInterface';

export class PercentageDiscountOffer implements OfferInterface {
  private product: Product;
  private unitPrice: number;
  private percentageOff: number;

  public constructor(product: Product, unitPrice: number, percentageOff: number) {
    this.product = product;
    this.unitPrice = unitPrice;
    this.percentageOff = percentageOff;
  }

  public getDiscount(cart: ShoppingCart): Discount {
    const quantity = cart.getQuantityOf(this.product);
    return new Discount(this.product, this.percentageOff + "% off", (quantity * this.unitPrice * this.percentageOff) / 100 );
  }

  public applies(cart: ShoppingCart): boolean {
    return cart.getQuantityOf(this.product) >= 1;
  }
}