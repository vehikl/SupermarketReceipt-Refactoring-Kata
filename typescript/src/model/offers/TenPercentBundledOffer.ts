import OfferInterface from './OfferInterface';
import { ShoppingCart } from '../ShoppingCart';
import { Discount } from '../Discount';
import { Product } from '../Product';
import { ProductUnit } from '../ProductUnit';
import { SupermarketCatalog } from '../SupermarketCatalog';

export class TenPercentBundledOffer implements OfferInterface {
  private bundledProducts: Array<Product>;
  private catalog: SupermarketCatalog;
  constructor(bundledProducts: Array<Product>, catalog: SupermarketCatalog) {
    this.bundledProducts = bundledProducts;
    this.catalog = catalog;
  }

  getDiscount(cart: ShoppingCart): Discount {
    const bundleCount = this.getBundleCount(cart);
    let totalPrice: number = this.getTotalPriceOfBundledProducts(bundleCount);
    return new Discount(new Product('product', ProductUnit.Kilo), '', totalPrice * 0.1);
  }

  private getTotalPriceOfBundledProducts(bundleCount: number) {
    let totalPrice: number = 0;
    this.bundledProducts.forEach(product => {
      totalPrice += this.catalog.getUnitPrice(product);
    });

    return totalPrice * bundleCount;
  }

  private getBundleCount(cart: ShoppingCart) {
    return this.bundledProducts
      .map(product => cart.getQuantityOf(product))
      .reduce((result, quantity) => Math.min(result, quantity));
  }

  applies(cart: ShoppingCart): boolean {
    return this.bundledProducts.every(product => {
      return cart.getQuantityOf(product) >= 1;
    });
  }
}
