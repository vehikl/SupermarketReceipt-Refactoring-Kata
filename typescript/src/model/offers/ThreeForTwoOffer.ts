import { Product } from "../Product"
import { Discount } from "../Discount"
import { ShoppingCart } from "../ShoppingCart";
import OfferInterface from './OfferInterface';
import DiscountInterface from "../DiscountInterface";

export class ThreeForTwoOffer implements OfferInterface {
  private minimumQuantityForOffer: number = 3;
  private product: Product;
  private unitPrice: number;

  public constructor(product: Product, unitPrice: number) {
    this.product = product;
    this.unitPrice = unitPrice;
  }

  public getDiscount(cart: ShoppingCart): DiscountInterface {
    const quantity = cart.getQuantityOf(this.product);
    const maybeDiscountMultiple = Math.floor(quantity / this.minimumQuantityForOffer);
    const discountAmount = quantity * this.unitPrice - ((maybeDiscountMultiple * 2 * this.unitPrice) + quantity % 3 * this.unitPrice);

    return new Discount(this.product, "3 for 2", discountAmount);
  }

  public applies(cart: ShoppingCart): boolean {
    return cart.getQuantityOf(this.product) > 2;
  }
}