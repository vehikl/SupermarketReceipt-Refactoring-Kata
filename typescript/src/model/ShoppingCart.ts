import { Product } from "./Product"
import { SupermarketCatalog } from "./SupermarketCatalog"
import * as _ from "lodash"
import { ProductQuantity } from "./ProductQuantity"
import { Discount } from "./Discount"
import { Receipt } from "./Receipt"
import { Offer } from "./Offer"
import { SpecialOfferType } from "./SpecialOfferType"
import { TwoForAmountOffer } from "./TwoForAmountOffer";
import { ThreeForTwoOffer } from "./ThreeForTwoOffer";
import { PercentageDiscountOffer } from "./PercentageDiscountOffer";
import { FiveForAmountOffer } from "./FiveForAmountOffer";
import OfferInterface from "./OfferInterface";

type ProductQuantities = { [productName: string]: ProductQuantity }
export type OffersByProduct = { [productName: string]: Offer };

export class ShoppingCart {

    private readonly items: ProductQuantity[] = [];
    _productQuantities: ProductQuantities = {};


    getItems(): ProductQuantity[] {
        return _.clone(this.items);
    }

    addItem(product: Product): void {
        this.addItemQuantity(product, 1.0);
    }

    productQuantities(): ProductQuantities {
        return this._productQuantities;
    }

    getQuantityOfNumber(product: Product): number {
        if (!this._productQuantities[product.name]) {
            return 0
        }
        return this._productQuantities[product.name].quantity;
    }


    public addItemQuantity(product: Product, quantity: number): void {
        let productQuantity = new ProductQuantity(product, quantity)
        this.items.push(productQuantity);
        let currentQuantity = this._productQuantities[product.name]
        if (currentQuantity) {
            this._productQuantities[product.name] = this.increaseQuantity(product, currentQuantity, quantity);
        } else {
            this._productQuantities[product.name] = productQuantity;
        }

    }

    private increaseQuantity(product: Product, productQuantity: ProductQuantity, quantity: number) {
        return new ProductQuantity(product, productQuantity.quantity + quantity)
    }

    handleOffers(receipt: Receipt, offers: OffersByProduct, catalog: SupermarketCatalog, allSpecialOffers: Array<OfferInterface>): void {

        if (allSpecialOffers && allSpecialOffers.length) {
            return this.calcDiscountUsingNewStructure(allSpecialOffers, receipt);
        }

        this.calcDiscountUsingTheOldWay(offers, catalog, receipt);
    }

    private calcDiscountUsingTheOldWay(offers: OffersByProduct, catalog: SupermarketCatalog, receipt: Receipt) {
        for (const productName in this.productQuantities()) {
            const productQuantity = this._productQuantities[productName];
            const product = productQuantity.product;
            if (offers[productName]) {
                const offer: Offer = offers[productName];
                const unitPrice: number = catalog.getUnitPrice(product);
                const offerType = offer.offerType;
                let discount = this.getDiscountForOffer(productName, offerType, offer, unitPrice, product);
                if (discount != null)
                    receipt.addDiscount(discount);
            }
        }
    }

    private calcDiscountUsingNewStructure(allSpecialOffers: OfferInterface[], receipt: Receipt) {
        allSpecialOffers.forEach(offer => {
            if (offer.applies(this)) {
                receipt.addDiscount(offer.getDiscount(this));
            }
        });
        return;
    }

    private getDiscountForOffer(productName: string, offerType: SpecialOfferType, offer: Offer, unitPrice: number, product: Product) {
        const quantity: number = this._productQuantities[productName].quantity;

        let discount: Discount | null = null;

        discount = this.getTwoForAmountDiscount(offerType, offer, unitPrice, discount, product);
        discount = this.threeForTwoDiscount(offerType, unitPrice, discount, product);
        discount = this.percentageDiscount(offerType, discount, product, offer, unitPrice);
        discount = this.fiveForAmountDiscount(offerType, unitPrice, offer, discount, product);
        return discount;
    }

    private fiveForAmountDiscount(offerType: SpecialOfferType, unitPrice: number, offer: Offer, discount: Discount | null, product: Product) {
        const fiveForAmount: FiveForAmountOffer = new FiveForAmountOffer(product, unitPrice, offer.argument);

        if (offerType == SpecialOfferType.FiveForAmount && fiveForAmount.applies(this)) {
            return fiveForAmount.getDiscount(this);
        }
        return discount;
    }

    private percentageDiscount(offerType: SpecialOfferType, discount: Discount | null, product: Product, offer: Offer, unitPrice: number) {
        const percentageDiscount: PercentageDiscountOffer = new PercentageDiscountOffer(product, unitPrice, offer.argument);
        if (offerType == SpecialOfferType.TenPercentDiscount && percentageDiscount.applies(this)) {
            return percentageDiscount.getDiscount(this);
        }
        return discount;
    }

    private threeForTwoDiscount(offerType: SpecialOfferType, unitPrice: number, discount: Discount | null, product: Product) {
        const threeForTwo: ThreeForTwoOffer = new ThreeForTwoOffer(product, unitPrice);

        if (offerType == SpecialOfferType.ThreeForTwo && threeForTwo.applies(this)) {
            return threeForTwo.getDiscount(this);
        }
        return discount;
    }

    private getTwoForAmountDiscount(offerType: SpecialOfferType, offer: Offer, unitPrice: number, discount: Discount | null, product: Product) {
        const twoForAmount: TwoForAmountOffer = new TwoForAmountOffer(product, unitPrice, offer.argument);

        if (offerType == SpecialOfferType.TwoForAmount && twoForAmount.applies(this)) {
            return twoForAmount.getDiscount(this);
        }
        return discount;
    }
}
