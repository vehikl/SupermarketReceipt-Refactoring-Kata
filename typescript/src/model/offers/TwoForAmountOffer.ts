import { Product } from "../Product"
import { Discount } from "../Discount"
import { ShoppingCart } from "../ShoppingCart";
import OfferInterface from './OfferInterface';

export class TwoForAmountOffer implements OfferInterface {
  private minimumQuantityForOffer: number = 2;
  private product: Product;
  private unitPrice: number;
  private discountedPrice: number;

  public constructor(product: Product, unitPrice: number, discountedPrice: number) {
    this.product = product;
    this.unitPrice = unitPrice;
    this.discountedPrice = discountedPrice;
  }

  public getDiscount(cart: ShoppingCart) {
    const quantity = cart.getQuantityOf(this.product);
    const total = this.discountedPrice * Math.floor(quantity / this.minimumQuantityForOffer) + quantity % 2 * this.unitPrice;
    const discountAmount = this.unitPrice * quantity - total;
    return new Discount(this.product, "2 for " + this.discountedPrice, discountAmount);
  }

  public applies(cart: ShoppingCart): boolean {
    return cart.getQuantityOf(this.product) >= 2;
  }
}