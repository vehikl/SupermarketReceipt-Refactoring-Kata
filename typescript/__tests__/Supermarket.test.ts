import { FakeCatalog } from "./FakeCatalog"
import { Product } from "../src/model/Product"
import { SupermarketCatalog } from "../src/model/SupermarketCatalog"
import { Receipt } from "../src/model/Receipt"
import { ShoppingCart } from "../src/model/ShoppingCart"
import { Teller } from "../src/model/Teller"
import { ProductUnit } from "../src/model/ProductUnit"
import { ReceiptPrinter } from "../src/ReceiptPrinter"
import { FiveForAmountOffer } from "../src/model/offers/FiveForAmountOffer";
import OfferInterface from "../src/model/offers/OfferInterface";
import { ThreeForTwoOffer } from "../src/model/offers/ThreeForTwoOffer";
import { PercentageDiscountOffer } from "../src/model/offers/PercentageDiscountOffer";
import { TwoForAmountOffer } from "../src/model/offers/TwoForAmountOffer";
import { TenPercentBundledOffer } from "../src/model/offers/TenPercentBundledOffer";

describe('Supermarket', () => {
    const applePrice: number = 1.99;
    const toothbrushPrice: number = 0.99;
    const ricePrice: number = 2.49;
    const toothpastePrice: number = 1.79;
    const cherryTomatoesPrice: number = 0.69;
    let toothbrush: Product;
    let toothpaste: Product;
    let apples: Product;
    let rice: Product;
    let cherryTomatoes: Product;
    let catalog: SupermarketCatalog;
    let cart: ShoppingCart;
    let teller: Teller;
    let receipt: Receipt;
    let printer: ReceiptPrinter;

    beforeEach(() => {
        catalog = new FakeCatalog();
        toothbrush = new Product("toothbrush", ProductUnit.Each);
        apples = new Product("apples", ProductUnit.Kilo);
        rice = new Product("rice", ProductUnit.Each);
        toothpaste = new Product("toothpaste", ProductUnit.Each);
        cherryTomatoes = new Product("cherry tomatoes", ProductUnit.Each);
        catalog.addProduct(toothbrush, toothbrushPrice);
        catalog.addProduct(apples, applePrice);
        catalog.addProduct(rice, ricePrice);
        catalog.addProduct(toothpaste, toothpastePrice);
        catalog.addProduct(cherryTomatoes, cherryTomatoesPrice);
        cart = new ShoppingCart();
        teller = new Teller(catalog);
        printer = new ReceiptPrinter();
    });

    it('applies three for two discount', () => {
        cart.addItemQuantity(toothbrush, 3);

        const threeForTwoOffer: OfferInterface = new ThreeForTwoOffer(toothbrush, toothbrushPrice);
        teller.addSpecialOffer(threeForTwoOffer);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toBeCloseTo(toothbrushPrice * 2);
        expect(printer.printReceipt(receipt)).toMatchSnapshot();
    });

    it('applies multiple offers', () => {
        const discountedPriceForTwo = 0.99;
        cart.addItemQuantity(toothbrush, 3);
        cart.addItemQuantity(cherryTomatoes, 3);

        const threeForTwoOffer: OfferInterface = new ThreeForTwoOffer(toothbrush, toothbrushPrice);
        const twoForOffer: OfferInterface = new TwoForAmountOffer(cherryTomatoes, cherryTomatoesPrice, discountedPriceForTwo);

        teller.addSpecialOffer(threeForTwoOffer);
        teller.addSpecialOffer(twoForOffer);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toBeCloseTo(toothbrushPrice * 2 + discountedPriceForTwo + cherryTomatoesPrice);
        expect(printer.printReceipt(receipt)).toMatchSnapshot();
    });

    it('applies twenty percent discount', () => {
        const discountPercentage = 20.0;
        const discountMultiplier = (100 - discountPercentage) / 100;
        cart.addItemQuantity(apples, 1);

        const percentageOffer: OfferInterface = new PercentageDiscountOffer(apples, applePrice, discountPercentage);
        teller.addSpecialOffer(percentageOffer);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toBeCloseTo(applePrice * discountMultiplier);
        expect(printer.printReceipt(receipt)).toMatchSnapshot();
    });

    it('applies ten percent discount', () => {
        const discountPercentage = 10.0;
        const discountMultiplier = (100 - discountPercentage) / 100;
        cart.addItemQuantity(rice, 1);

        const percentageOffer: OfferInterface = new PercentageDiscountOffer(rice, ricePrice, discountPercentage);
        teller.addSpecialOffer(percentageOffer);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toBeCloseTo(ricePrice * discountMultiplier);
        expect(printer.printReceipt(receipt)).toMatchSnapshot();
    });

    it('applies five for amount discount', () => {
        const discountedPriceForFive: number = 7.49;
        cart.addItemQuantity(toothpaste, 5);

        const offer: OfferInterface = new FiveForAmountOffer(toothpaste, toothpastePrice, discountedPriceForFive);
        teller.addSpecialOffer(offer);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toEqual(discountedPriceForFive);
        expect(printer.printReceipt(receipt)).toMatchSnapshot();
    });

    it('applies two for amount discount', () => {
        const discountedPriceForTwo: number = 0.99;
        cart.addItem(cherryTomatoes);
        cart.addItem(cherryTomatoes);

        const offer: OfferInterface = new TwoForAmountOffer(cherryTomatoes, cherryTomatoesPrice, discountedPriceForTwo);
        teller.addSpecialOffer(offer);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toEqual(discountedPriceForTwo);
        expect(printer.printReceipt(receipt)).toMatchSnapshot();
    });

    it('does not apply two for amount if 1 item purchased', () => {
        const discountedPriceForTwo: number = 0.99;
        cart.addItem(cherryTomatoes);

        const offer: OfferInterface = new TwoForAmountOffer(cherryTomatoes, cherryTomatoesPrice, discountedPriceForTwo);
        teller.addSpecialOffer(offer);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toEqual(cherryTomatoesPrice);
        expect(printer.printReceipt(receipt)).toMatchSnapshot();
    });

    it('does not apply discount to items not purchased', () => {
        cart.addItem(toothpaste);

        const offer: OfferInterface = new TwoForAmountOffer(cherryTomatoes, cherryTomatoesPrice, 0.99);
        teller.addSpecialOffer(offer);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toEqual(toothpastePrice);
        expect(printer.printReceipt(receipt)).toMatchSnapshot();
    });

    it('applies a bundle discount with one of each in the cart', () => {
        const bundledOffer: OfferInterface = new TenPercentBundledOffer([toothbrush, toothpaste], catalog);
        teller.addSpecialOffer(bundledOffer);

        cart.addItem(toothbrush);
        cart.addItem(toothpaste);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toBeCloseTo((toothbrushPrice + toothpastePrice) * 0.9);
    });

    it('applies a bundle discount with another one of each in the cart', () => {
        const bundledOffer: OfferInterface = new TenPercentBundledOffer([cherryTomatoes, apples], catalog);
        teller.addSpecialOffer(bundledOffer);

        cart.addItem(cherryTomatoes);
        cart.addItem(apples);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toBeCloseTo((cherryTomatoesPrice + applePrice) * 0.9);
    });

    it('only applies bundle discounts to applicable products', () => {
        const bundledOffer: OfferInterface = new TenPercentBundledOffer([cherryTomatoes, apples], catalog);
        teller.addSpecialOffer(bundledOffer);

        cart.addItem(toothbrush);

        receipt = teller.checksOutArticlesFrom(cart);

        expect(receipt.getTotalPrice()).toBeCloseTo(toothbrushPrice);
    });

    it('applies a bundle discount to a pair of bundled products', () => {
        const bundledOffer: OfferInterface = new TenPercentBundledOffer([cherryTomatoes, apples], catalog);
        teller.addSpecialOffer(bundledOffer);

        cart.addItemQuantity(cherryTomatoes, 2);
        cart.addItemQuantity(apples, 2);

        receipt = teller.checksOutArticlesFrom(cart);

        const discountedPriceForOneBundle = (cherryTomatoesPrice + applePrice) * 0.9;
        expect(receipt.getTotalPrice()).toBeCloseTo(discountedPriceForOneBundle * 2);
        expect(printer.printReceipt(receipt)).toMatchSnapshot();
    });
});
