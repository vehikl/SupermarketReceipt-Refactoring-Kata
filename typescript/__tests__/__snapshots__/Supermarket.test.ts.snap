// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Supermarket applies a bundle discount to a pair of bundled products 1`] = `
"cherry tomatoes                     1.38
  0.69 * 2
apples                              3.98
  1.99 * 2.000
10% off(cherry tomatoes & apples)  -0.54

Total:                              4.82"
`;

exports[`Supermarket applies five for amount discount 1`] = `
"toothpaste                          8.95
  1.79 * 5
5 for 7.49(toothpaste)             -1.46

Total:                              7.49"
`;

exports[`Supermarket applies multiple offers 1`] = `
"toothbrush                          2.97
  0.99 * 3
cherry tomatoes                     2.07
  0.69 * 3
3 for 2(toothbrush)                -0.99
2 for 0.99(cherry tomatoes)        -0.39

Total:                              3.66"
`;

exports[`Supermarket applies ten percent discount 1`] = `
"rice                                2.49
10% off(rice)                      -0.25

Total:                              2.24"
`;

exports[`Supermarket applies three for two discount 1`] = `
"toothbrush                          2.97
  0.99 * 3
3 for 2(toothbrush)                -0.99

Total:                              1.98"
`;

exports[`Supermarket applies twenty percent discount 1`] = `
"apples                              1.99
20% off(apples)                    -0.40

Total:                              1.59"
`;

exports[`Supermarket applies two for amount discount 1`] = `
"cherry tomatoes                     0.69
cherry tomatoes                     0.69
2 for 0.99(cherry tomatoes)        -0.39

Total:                              0.99"
`;

exports[`Supermarket does not apply discount to items not purchased 1`] = `
"toothpaste                          1.79

Total:                              1.79"
`;

exports[`Supermarket does not apply two for amount if 1 item purchased 1`] = `
"cherry tomatoes                     0.69

Total:                              0.69"
`;
